### PR TITLE
Fix #18055 - Add null annotation to Async.SwitchToContext

### LINKED ISSUE #18055: Nullness issue - Async.SwitchToContext lacks a null annotation on its parameter
### Issue description

I believe that Async.SwitchToContext should take a parameter of type `System.Threading.SynchronizationContext | null`. Examining the function body, you can see that it treats null as a possible argument and handles it normally (without throwing an exception).

### Choose one or more from the following categories of impact

- [x] Unexpected nullness warning (false positive in nullness checking, code uses --checknulls and langversion:preview).
- [ ] Missing nullness warning in a case which can produce nulls (false negative, code uses --checknulls and langversion:preview).
- [ ] Breaking change related to older `null` constructs in code not using the checknulls switch.
- [ ] Breaking change related to generic code and explicit type constraints (`null`, `not null`).
- [ ] Type inference issue (i.e. code worked without type annotations before, and applying the --checknulls enforces type annotations).
- [ ] C#/F# interop issue related to nullness metadata.
- [ ] Other (none of the categories above apply).

### Operating System

Windows (Default)

### What .NET runtime/SDK kind are you seeing the issue on

.NET SDK (.NET Core, .NET 5+)

### .NET Runtime/SDK version

net9.0

### Reproducible code snippet and actual behavior

_No response_

### Possible workarounds

One can write a wrapper function which when given null calls Async.SwitchToThreadPool, otherwise calls SwitchToContext with the non-null syncContext.

### PR DESCRIPTION
**Pull request by BoundedChenn31**
## Description

Fixes #18055

## Checklist

- [ ] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [ ] Release notes entry updated:
    > Please make sure to add an entry with short succinct description of the change as well as link to this pull request to the respective release notes file, if applicable.
    >
    > Release notes files:
    > - If anything under `src/Compiler` has been changed, please make sure to make an entry in `docs/release-notes/.FSharp.Compiler.Service/<version>.md`, where `<version>` is usually "highest" one, e.g. `42.8.200`
    > - If language feature was added (i.e. `LanguageFeatures.fsi` was changed), please add it to `docs/release-notes/.Language/preview.md`
    > - If a change to `FSharp.Core` was made, please make sure to edit `docs/release-notes/.FSharp.Core/<version>.md` where version is "highest" one, e.g. `8.0.200`.

    > Information about the release notes entries format can be found in the [documentation](https://fsharp.github.io/fsharp-compiler-docs/release-notes/About.html).
    > Example:
    > * More inlines for Result module. ([PR #16106](https://github.com/dotnet/fsharp/pull/16106))
    > * Correctly handle assembly imports with public key token of 0 length. ([Issue #16359](https://github.com/dotnet/fsharp/issues/16359), [PR #16363](https://github.com/dotnet/fsharp/pull/16363))
    > *`while!` ([Language suggestion #1038](https://github.com/fsharp/fslang-suggestions/issues/1038), [PR #14238](https://github.com/dotnet/fsharp/pull/14238))

    > **If you believe that release notes are not necessary for this PR, please add `NO_RELEASE_NOTES` label to the pull request.**

### CODE REVIEW

**Comment by BoundedChenn31** [comment]
@dotnet-policy-service agree

**Comment by BoundedChenn31** [comment]
I thinл that release notes are not necessary but unable to add corresponding label :)

**Review by T-Gro — APPROVED** [review]

File: src/FSharp.Core/async.fs [hunk 1]
```diff
@@ -868,11 +868,7 @@ module AsyncPrimitives =
 
     ///   - Initial cancellation check
     ///   - Call syncCtxt.Post with exception protection. This may fail as it is arbitrary user code
-#if BUILDING_WITH_LKG || NO_NULLCHECKING_LIB_SUPPORT
```
Comment (T-Gro) [inline src/FSharp.Core/async.fs:871]: Thank you for this contribution.
This reminds me that with .NET9 release, we do not need most (maybe all?) of these  directives anymore - those were mainly needed when the old compiler used for building this compiler did not understand the new syntax.

There might still be situations for nullness-related bugs where we would want to build the compiler with nullness checking disabled (e.g. if there is a regression not allowing it any other way), but the syntax should be just fine.

**Review by psfinaki — APPROVED** [review]
Thanks for your first contribution, @BoundedChenn31! Way to go!

File: src/FSharp.Core/async.fs [hunk 2]
```diff
@@ -868,11 +868,7 @@ module AsyncPrimitives =
...
     let CreateSwitchToAsync (syncCtxt: SynchronizationContext) =
```
Comment (vzarytovskii) [inline src/FSharp.Core/async.fs:871]: shall fsi and fs be synchromised?

Comment (T-Gro) [reply src/FSharp.Core/async.fs:871]: This method is not part of the public signature. its in async.fs only.

**Review by edgarfgp — APPROVED** [review]
Thanks @BoundedChenn31. It is great to see new people contributing here.

In AmplifyingF# we have several sessions about working on the compiler and other F# OSS projects. In case you are interested in knowing more.

Edit: https://amplifyingfsharp.io/sessions/

**Comment by psfinaki** [comment]
@kerams thanks for chiming in, it's fair that the note can be beneficial here - I added it.

**Pull request merged**